### ISSUE TITLE
Using the provided eval code on different dataset

### ISSUE THREAD

**Issue by salokr** [ElementAI/picard#36]
Hey, 

I have tried running "make eval" and was able to retrieve numbers on Spider-Dev. 

Now, I would like to use the provided T5+PICARD for evaluation on other datasets like Spider-DK or Spider-realistic. Is there an easy and convenient way to use these dataset for eval only---for e.g., as a command-line arguments?

Thanks

**Comment by tscholak** [comment]
Hi @salokr,
I'm afraid there isn't a code path for these datasets. Contributions are welcome. You'd have to make a copy of https://github.com/ElementAI/picard/blob/main/seq2seq/datasets/spider/spider.py, e.g. https://github.com/ElementAI/picard/blob/main/seq2seq/datasets/spider-realistic/spider-realistic.py, and then change it so that it downloads the dataset and pre-processes it correctly. You also need to add some straight-forward code here: https://github.com/ElementAI/picard/blob/e37020b6eee18bff865d9d2ba852bd636f3ed777/seq2seq/utils/dataset_loader.py#L86.
Good luck, Torsten

**Comment by tscholak** [comment]
see also #59 #58 #57

**Comment by salokr** [comment]
Thank you for this @tscholak .

I have one more query. Can I avoid using the docker images altogether and use the code as it is. Is there a way I can do this?

I will train T5 on my own dataset and then will follow the steps to eval and do the rest of the things. I tried installing packages from the ".toml" file using "python -m pip install ." but I'm getting an error that "PICARD is not a package" so the installation procedure halts every time.

Saurabh

**Comment by tscholak** [comment]
Hi @salokr,
Yes, you can train your model with just the python code and dependencies. The repository uses poetry, and you can use `poetry install` to install the python dependencies. This should not give you an error like the above.
Contributions for additional datasets are welcome!

**Comment by salokr** [comment]
Hi @tscholak thank you for the response. I tried using "poetry install" but the same error "picard in not a package" is there. Attaching the screenshot for your reference:

![image](https://user-images.githubusercontent.com/19395011/154393373-9b26230f-00a6-4c92-a6af-0870f2da3943.png)




`Contributions for additional datasets are welcome!`:
I can give it a try but with no promises :)



Saurabh

**Comment by tscholak** [comment]
🤔 perhaps `poetry install --no-root` will work. This should only install dependencies.

**Comment by salokr** [comment]
I think, in the last run `poetry install` packages were installed. Because when I tried running `poetry install --no-root`, I got:

![image](https://user-images.githubusercontent.com/19395011/154397160-b80df765-f790-48b6-9a02-390346a1e4ba.png)

**Comment by salokr** [comment]
Could you please suggest to me steps that I should follow to train my model on some dataset, say spider, then evaluate it on some other dataset, like spider_syn `without using the docker images?

Saurabh

**Comment by salokr** [comment]
Hi @tscholak ,

I tried Spider-Real with PICARD here are the number:

![image](https://user-images.githubusercontent.com/19395011/155032637-2297615b-84c1-4889-8013-32f04d07f88f.png)

Saurabh

**Comment by tscholak** [comment]
Hey @salokr,
Thanks for sharing these numbers, they are quite interesting. Is there any code you could share? There is an open issue and call for help with implementing additional dataset code. It would be amazing if you were to contribute by opening a pull request. 😊

**Comment by salokr** [comment]
Hi @tscholak,

I have added support for Spider-Syn ( I see that somebody has already finished for real ). I was waiting for the [bug](https://github.com/ElementAI/picard/pull/github.com/ElementAI/picard/issues/61) to be resolved but now I guess we can resolve this by bypassing the local reference to URL_ variable for the datasets as suggested by [.](https://github.com/ElementAI/picard/pull/69)

Here are my number on all versions of Spider:

<img width="851" alt="Screen Shot 2022-03-04 at 2 09 06 PM" src="https://user-images.githubusercontent.com/19395011/156826492-4038b0b9-2f61-4844-a99c-6de353bcce62.png">

**Comment by tscholak** [comment]
thank you @salokr!

**Comment by tscholak** [comment]
Hi @salokr,
The issue I would have with that is that someone would need to maintain this. I am unfamiliar with Singularity, and I therefore cannot fix any issue that may come up.

**Comment by tscholak** [comment]
That being said, I'm curious how the procedure would look like. If it's just another section in the readme, I'm probably ok with it :)

**Comment by salokr** [comment]
Upon reading [here](https://www.nas.nasa.gov/hecc/support/kb/best-practices-for-running-singularity-on-nas-systems_659.html), it turns out that there is a simpler solution to edit docker images, and it doesn't even require pushing the changes back to git (like in docker) --- using the singularity and the sandbox.

Using a sandbox we can create a container within a writable directory. The resulting directory operates just like a container in a SIF/docker file but is *editable*. 

In this way, you can navigate to the app/​ directory of this editable container and make changes directly to the files (no need to push to git to reflect changes), and simply execute the seq2seq/run_seq2seq.py file.

In summary, it will be like using a docker image within our system but with edit permissions. Singularity creates a writable directory called `app` where all the files and folders from the repo are stored. Edit them and directly run. Probably, there is a total of 10 steps (7 for singularity and 3 to make changes to add support for spider_real/dk/dev) to get eval numbers on a new dataset.

**Comment by tscholak** [comment]
Thanks for this information.
What are you going to do about the non-python dependencies?

**Comment by salokr** [comment]
Whatever you have provided in the container will be there no need to install anything from the external sources. You just have to execute 

`singularity build --sandbox text-to-sql-eval docker://tscholak/text-to-sql-eval:e37020b6eee18bff865d9d2ba852bd636f3ed777         # using the --sandbox flag will create a writable directory text-to-sql-eval instead of a non-editable "sif" file. It will take time though (like 15-20 minutes)`

Singularity will create a directory called `text-to-sql-eval `. Within this directory all the environments and packages (conda etc.) you've provided all already available. Something like this:
![image](https://user-images.githubusercontent.com/19395011/157131076-d102b44a-f54d-4fea-b727-8634b866619f.png)

Upon navigating to ./opt/conda/bin, you will find:
![image](https://user-images.githubusercontent.com/19395011/157131242-9894c025-c592-4637-b57c-935a35dde0f7.png)

And inside ./app directory all the files related to this git are there:
![image](https://user-images.githubusercontent.com/19395011/157131341-60a3febd-549a-4763-a820-1b4b3c58d5ac.png)

Finally, make the desired changes inside any of the files you want and run the image (like you run docker).

**Comment by salokr** [comment]
It may seem complex a bit but I have found this solution to be easier (this is how I got the numbers on all versions of spider).

**Comment by tscholak** [comment]
Thanks again!
Have you confirmed that the binaries actually work inside this environment? I imagine there will be issues with dynamic linking of libraries.

**Comment by salokr** [comment]
I can take a look into this if you can please provide me to test which/how part of "the binaries actually work inside this environment?".

My experience with this is: I have used your eval image and used the following steps to get spider numbers (nothing additional) and I never faced any issue regarding libraries/dependencies :

![image](https://user-images.githubusercontent.com/19395011/157132952-9cc1c3f8-d6aa-4144-be7c-d6098e2c6682.png)

**Comment by tscholak** [comment]
Alright then, feel free to add a section to the readme!

**Comment by salokr** [comment]
Sure thanks :)
